### PR TITLE
Add object version-ID support

### PR DESCRIPTION
**Pull request by balamurugana**
Added support to below APIs
* enable_bucket_versioning()
* disable_bucket_versioning()
* fget_object()
* get_object()
* stat_object()
* remove_object()
* remove_objects()
* presigned_get_object()
* list_objects()
* list_objects_v2()

### CODE REVIEW

File: minio/api.py [hunk 1]
```diff
@@ -1117,34 +1127,49 @@ def stat_object(self, bucket_name, object_name, sse=None):
...
-        return Object(bucket_name, object_name, last_modified, etag, size,
-                      content_type=content_type, metadata=custom_metadata)
 
-    def remove_object(self, bucket_name, object_name):
+        return Object(
+            bucket_name,
+            object_name,
+            last_modified=last_modified,
+            etag=response.headers.get("etag", "").replace('"', ""),
+            size=int(response.headers.get("content-length", "0")),
```
Comment (Praveenrajmani) [inline minio/api.py:1166]: I see these headers being canonicalized in s3 docs. Like  `Content-Length` and `Content-Type` etc. Are we cleaning it any place?

looks like all the x-amz-* will be non-canonicalized as per the docs - https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadObject.html

File: minio/api.py [hunk 2]
```diff
@@ -312,6 +314,7 @@ def make_bucket(self, bucket_name, location='us-east-1'):
         Examples:
             minio.make_bucket('foo')
             minio.make_bucket('foo', 'us-west-1')
+            minio.make_bucket('foo', 'us-west-1', object_lock=True)
 
         :param bucket_name: Bucket to create on server
         :param location: Location to create bucket on
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply minio/api.py:320]: done

File: minio/api.py [hunk 3]
```diff
@@ -689,7 +725,8 @@ def fput_object(self, bucket_name, object_name, file_path,
                                    progress, part_size)
 
     def fget_object(self, bucket_name, object_name, file_path,
-                    request_headers=None, sse=None):
+                    request_headers=None, sse=None, version_id=None,
+                    extra_query_params=None):
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply minio/api.py:730]: done

File: minio/api.py [hunk 4]
```diff
@@ -961,7 +1016,8 @@ def put_object(self, bucket_name, object_name, data, length,
                                    metadata=metadata, sse=sse,
                                    progress=progress)
 
-    def list_objects(self, bucket_name, prefix='', recursive=False):
+    def list_objects(self, bucket_name, prefix=None, recursive=False,
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply minio/api.py:1026]: I think it is redundant to have one more API for versioning (in fact `list_objects_v2()` is also redundant and needs to be removed).  Many calls are not useful.

File: minio/api.py (hunk 1 above)
Comment (balamurugana) [reply minio/api.py:1173]: I think its using `FoldCaseDict` internally. Nothing changed from existing logic.

File: minio/api.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Praveenrajmani) [reply minio/api.py:1026]: Ok.. Then can we have `version_id_marker` support as well?

[the remark this answers is not in the document]
Comment (balamurugana) [reply minio/api.py:1026]: currently we expose very limited parameters in listing (`prefix`/fixed delimiter `/` via recursive arg in list_objects() and `prefix`/fixed delimiter `/` by recursive arg/`start_after` in list_objects_v2()). However we would need to expose few more useful args to have controlled listing.
This PR is to address versioning support rather specific to `ListObjects`. I will be add those useful args (`version_id_marker` is one such arg) later.

[the remark this answers is not in the document]
Comment (Praveenrajmani) [reply minio/api.py:1026]: Ok

**Comment by harshavardhana** [comment]
The functional tests are failing against AWS S3 @balamurugana looks like versions are not cleaned up properly during tear down for each versioning tests. 

NOTE: To delete a version versionId is necessary without that bucket cannot be emptied. PTAL

**Comment by harshavardhana** [comment]
After adding those changes was able to successfully test this against the versioning PR 

```
{"name": "minio-py:test_stat_object_version", "status": "PASS", "args": {"bucket_name": "minio-py-test-c34be9ba-fab3-473f-85ad-724c30646aa0", "object_name": "e4875001-bb44-4699-a3c4-9208350d235f-metadata", "length": 11534336, "data": "LimitedRandomReader(11 * MB)", "metadata": {"X-Amz-Meta-Testing": "value"}, "content_type": "application/octet-stream"}, "duration": 5364}
{"name": "minio-py:test_stat_object_version_SSEC", "status": "PASS", "args": {"bucket_name": "minio-py-test-ec8b21ea-0f65-43a3-9227-f9988610d3da", "object_name": "f99860c5-320d-4783-b4c6-0919893d5c46-metadata", "length": 11534336, "data": "LimitedRandomReader(11 * MB)", "metadata": {"X-Amz-Meta-Testing": "value"}, "content_type": "application/octet-stream"}, "duration": 4834}
{"name": "minio-py:test_get_object_version", "status": "PASS", "args": {"bucket_name": "minio-py-test-c15c2ce8-8280-42e4-9297-8055ce61c75c", "object_name": "134cff45-f869-48e6-80fb-3138a524f5db"}, "duration": 489}
{"name": "minio-py:test_get_object_version_SSEC", "status": "PASS", "args": {"bucket_name": "minio-py-test-e2cce067-7363-4ef2-9a45-9d56fef3bb49", "object_name": "09e7ea33-40d3-4159-8167-d58d4079e867"}, "duration": 573}
```

**Comment by balamurugana** [comment]
> The functional tests are failing against AWS S3 @balamurugana looks like versions are not cleaned up properly during tear down for each versioning tests.
> 
> NOTE: To delete a version versionId is necessary without that bucket cannot be emptied. PTAL

I haven't run functional tests against AWS, but tested individual APIs. I will fix the tests.

**Comment by balamurugana** [comment]
Fixed functional tests to handle Version-ID in all `finally`.

**Comment by kannappanr** [comment]
ping @Praveenrajmani @vadmeste

**Comment by vadmeste** [comment]
@balamurugana I have a doubt about something. Maybe we need to implement a different API for versioning listing, the reason is that users could be confused since versioning listing does not return object's type and its metadata

**Comment by harshavardhana** [comment]
> @balamurugana I have a doubt about something. Maybe we need to implement a different API for versioning listing, the reason is that users could be confused since versioning listing does not return object's type and its metadata

We will have an API on the server to facilitate metadata with versioning as well @vadmeste

**Comment by harshavardhana** [comment]
@balamurugana PTAL at the build failures

**Pull request merged**